ARROW-7096: [C++] Unified ConcatenateTables APIs

brills opened the pull request:
* Introduced a UnifySchemasOptions.
  - Currently the only option is to control whether to promote a Null
    column to a non-Null. (If turned off, UnifySchemas will only union the
    columns and raise an error on any mismatched types).
  - We can introduce more switches to control whether to promote
    integers, floats, etc (through casting).
  - We can also allow strategy of metadata merging to be specified.
* Introduced a ConcatenateTablesOptions.
  - Currently the only option is to control whether to unify schemas.
* For now PromoteTableToSchema won't accept any options. I'm not sure if
  this is desirable. But my vision is to make it do the best effort to
  cast the columns to the types specified in the desired schema. It's
  the caller's responsibility to make sure the desired schema is "safe"
  (or not).

Tests have not been updated fully. Test cases need to be added for UnifySchemas where promote_null == false. The ConcatenateTables test cases and ContenateTablesWithPromotion  test caes may need to be merged (i.e. using the same test fixture).

The python APIs / tests have not been changed.  (do we want to introduce the counter-part of those options class in the python API?)

fsaintjacques commented:
See https://github.com/apache/arrow/pull/6056 as it is related (and conflicting merge).

brills commented:
incorporated #6056 by @fsaintjacques :

- Introduced a `Field::MergeOptions` to control the behavior of `Field::MergeWith()`. 
  * For now, there is only one option to control whether to promote a Null type to a non-Null type (if false, the two fields must be of the same type).
  * options to control type casting can be added later here.
- the c'tor of `SchemaBuilder` will take `Field::MergeOptions`. (only effective with policy MERGE).
- Again, `ConcatenateTables()` and `ConcatenateTablesWithPromotion()` are unified into one `ConcatenateTables` that takes a option struct.

One implication that I think needs discussion is: this will lead to `PromoteTableToSchema()` doing the best effort eventually (i.e. if field 'a' of the target schema is of a type that is "compatible" in any sense (null, wider integer, etc), it will cast / promote without leaving the caller an option to disallow such behavior. Note sure if we need yet another set of switches for `PromoteTableToSchema()`.
 
Tests are not updated yet. New test cases are needed and will be added once the overall approach is LGTM'ed.

fsaintjacques commented on c_glib/arrow-glib/table.cpp:
```diff
@@ -553,9 +553,9 @@ garrow_table_concatenate(GArrowTable *table,
-  auto status = arrow::ConcatenateTables(arrow_tables, &arrow_concatenated_table);
-  if (garrow_error_check(error, status, "[table][concatenate]")) {
-    return garrow_table_new_raw(&arrow_concatenated_table);
+  auto arrow_concatenated_table = arrow::ConcatenateTables(arrow_tables);
```
Variable shadowing.

kou replied:
@brills Could you remove `std::shared_ptr<arrow::Table> arrow_concatenated_table;`?

fsaintjacques commented on cpp/src/arrow/type.h:
```diff
@@ -358,21 +358,33 @@ class ARROW_EXPORT Field : public detail::Fingerprintable {
+    /// The unified field will be of the other type and become nullable.
+    /// Nullability will be promoted to the looser option (nullable if one is not
+    /// nullable).
+    bool promote_null_type = true;
```
Maybe rename to `promote_nullability` since it also promote when none of the fields are of NullType.

bkietz commented on c_glib/arrow-glib/table.cpp:
```diff
@@ -553,9 +553,9 @@ garrow_table_concatenate(GArrowTable *table,
-  if (garrow_error_check(error, status, "[table][concatenate]")) {
-    return garrow_table_new_raw(&arrow_concatenated_table);
+  auto arrow_concatenated_table = arrow::ConcatenateTables(arrow_tables);
+  if (garrow::check(error, arrow_concatenated_table, "[table][concatenate]")) {
```
This looks like an opportunity for a `Result` helper function:
```suggestion
  if (auto t = garrow::check_moved(error, std::move(arrow_concatenated_table), "[table][concatenate]")) {
    return garrow_table_new_raw(&t);
```

kou replied:
That's interesting but I think that this is out of scope of this pull request.
We can consider this approach in a separated issue.

Cc: @shiro615 @mrkn

mrkn replied:
It's very interesting to me, too.  I've create [the new ticket](https://issues.apache.org/jira/browse/ARROW-7504) to do it.

brills replied:
Mark as resolved (out of scope).

wesm commented on cpp/src/arrow/table.cc:
```diff
@@ -522,61 +522,59 @@ Status Table::RenameColumns(const std::vector<std::string>& names,
-                         std::shared_ptr<Table>* table) {
+Result<std::shared_ptr<Table>> ConcatenateTables(
+    const std::vector<std::shared_ptr<Table>>& tables,
+    const ConcatenateTablesOptions options, MemoryPool* memory_pool) {
```
This `const` doesn't seem to be here in `table.h`

brills replied:
did you mean the const before ConcatenateTablesOptions?
it's intended as marking a value type (as oppose to reference / pointer) as const has not effect in a function declaration.

wesm commented:
seems like it would be good to get this into 0.16.0. I can help with reviewing, but can others assist with fixing the small issues with the build and rebasing?

brills commented:
Sorry I haven't got enough cycles since the first revision. I'll make some progress by this weekend.

brills commented:
I fixed the build (at least glib, parquet and python can build locally now).

Still working on tests.

brills commented:
Added tests for Field::MergeWith.

Still need to refactor tests of ConcatenateTables to reflect the API change.

Do we want to change the python concat_tables API in this PR?

wesm commented:
Thanks @brills I will review

wesm commented on cpp/src/arrow/type_test.cc:
```diff
@@ -227,6 +227,66 @@ TEST(TestField, TestReplacement) {
+    auto f2 = field("f", int32())->WithNullable(true);
+    std::shared_ptr<Field> result;
+    ASSERT_OK_AND_ASSIGN(result, f1->MergeWith(f2, options));
+    std::cout << "XXX: " << f1->ToString();
```
This looks like a debugging remnant

wesm commented:
Merging this. Thank you @brills! If there is follow up work to do in Python please open a new JIRA
